### PR TITLE
Add membership setting function

### PR DESCRIPTION
**Pull request by jefferai**
This adds the function that, given a list of managed groups and an account, handles updating memberships.

### CODE REVIEW

File: internal/auth/oidc/repository_managed_group_members.go [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+	"github.com/hashicorp/boundary/internal/kms"
+	"github.com/hashicorp/boundary/internal/oplog"
+)
+
+// SetManagedGroupMemberships will set the managed groups for the given account
+// ID. If mgs is empty, the set of groups the account belongs to will be
+// cleared. It returns the set of managed group IDs.
+//
+// mgs contains the set of managed groups that matched. It must contain the
+// group's version as this is used to ensure consistency.
```
Comment (jimlambrt) [inline internal/auth/oidc/repository_managed_group_members.go:18]: Perhaps add: "it's understood that managed groups are eventually consistent and that these msgs represent the point in time we will consist the msgs to."

File: internal/auth/oidc/repository_managed_group_members.go [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+				}
+				if rowsUpdated != 1 {
+					return errors.New(errors.MultipleRecords, op, fmt.Sprintf("updated oidc managed group and %d rows updated", rowsUpdated))
+				}
+				msgs = append(msgs, &mgOplogMsg)
+			}
+
+			currentMemberships, err = r.ListManagedGroupMembershipsByMember(ctx, acct.PublicId, WithReader(reader))
+			if err != nil {
+				return errors.Wrap(err, op, errors.WithMsg("unable to retrieve current principal roles after sets"))
```
Comment (jimlambrt) [inline internal/auth/oidc/repository_managed_group_members.go:111]: need to update WithMsg(...)

File: internal/auth/oidc/repository_managed_group_members.go [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+
+			if len(msgs) > 0 {
+				if err := w.WriteOplogEntryWith(ctx, oplogWrapper, mgTicket, metadata, msgs); err != nil {
+					return errors.Wrap(err, op, errors.WithMsg("unable to write oplog"))
+				}
+			}
+
+			currentMemberships, err = r.ListManagedGroupMembershipsByMember(ctx, acct.PublicId, WithReader(reader))
+			if err != nil {
+				return errors.Wrap(err, op, errors.WithMsg("unable to retrieve current principal roles after sets"))
```
Comment (jimlambrt) [inline internal/auth/oidc/repository_managed_group_members.go:181]: need to update this WithMsg as well.

File: internal/auth/oidc/repository_managed_group_members_test.go [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+
+func Test_ManagedGroupMemberships(t *testing.T) {
+	// This tests both managed group membership functions (set/list) as list is
+	// always called as a return from set and we are validating the values that
+	// come back against what we expect.
+
+	// This test can be run in parallel; the subtests *cannot*.
+	t.Parallel()
+
+	// Note: using a test controller here for ease of setup as we need a working
```
Comment (jimlambrt) [inline internal/auth/oidc/repository_managed_group_members_test.go:30]: ty for the comment!

File: internal/auth/oidc/repository_managed_group_members_test.go [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+		mg.Filter = testFakeManagedGroupFilter
+		got, err := repo.CreateManagedGroup(context.Background(), org.GetPublicId(), mg)
+		require.NoError(t, err)
+		mgs = append(mgs, got)
+	}
+
+	// Fetch valid OIDC accounts. One will be "static" where we will simply
+	// ensure modifying the groups for the other doesn't affect it; the other
+	// will be used for testing.
+	rows, err := rw.Query(ctx, "select public_id from auth_oidc_account limit 2", nil)
```
Comment (jimlambrt) [inline internal/auth/oidc/repository_managed_group_members_test.go:74]: you could refactor this to just use vanilla db.SearchWhere(...) WithLimit(2) and an []Account

File: internal/auth/oidc/repository_managed_group_members_test.go [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+)
+
+const testFakeManagedGroupFilter = `"/foo" == "bar"`
+
+func Test_ManagedGroupMemberships(t *testing.T) {
+	// This tests both managed group membership functions (set/list) as list is
+	// always called as a return from set and we are validating the values that
+	// come back against what we expect.
+
+	// This test can be run in parallel; the subtests *cannot*.
```
Comment (jimlambrt) [inline internal/auth/oidc/repository_managed_group_members_test.go:27]: ty for the comment!

**Review by jimlambrt — CHANGES_REQUESTED** [review]
Great work and the tests are really well done.  Just a few minor cut/paste bits.

**Comment by jefferai** [comment]
@jimlambrt I've addressed all feedback in 61637ca

**Review by jimlambrt — APPROVED** [review]

**Review by vancluever — APPROVED** [review]
LGTM, again, mainly informational for me, but also again great context for my understanding of the codebase 🙂

**Pull request merged**